RIOT-OS/RIOT
# tools/mcuboot/imgtool: implement deterministic signatures.

### Contribution description

The behavior of imgtool is not deterministic:

- Each it is run it generates a different key.
- Even with the same key, each signature operation produces different
  results.

This is expected, provides security, and the correct behaviour of the cryptographic standards used.

The problem, however, is that in a CI environment one usually wants reproducible results.

This commit adds a new option to the `imgtool` (--seed) that replaces the default secure RNG with an __insecure__ PRNG initialized with a user-supplied seed. It also toggles deterministic signing in ECDSA.

This option sould NEVER, EVER be used in a production environment.

The mcuboot test is modified to detect when it is running in the CI and tells `imgtool` to use deterministic key generation and signing with a predefined seed. This will allows the test to produce the exact same binary in each CI run.

### Testing procedure

This should be tested with and without the last PR (last pr=ECDSA, without last pr = RSA).

If you use make (or the tool) normally in this branch you get the default behaviour, but you don't have to believe me and you can checkout master too. The new feature is enabled in the build system either by setting `RIOT_CI_BUILD` or `IMGTOOL_DETERMINISTIC`

Go to the `tests/mcuboot` directory. First we test key generation:

```sh
$ make clean 
$ make mcuboot-create-key 
$ md5sum bin/nrf52dk/key.pem 
271d0b9bb8b96b0705b450b8a27ba213  bin/nrf52dk/key.pem
$ make clean 
$ make mcuboot-create-key 
$ md5sum bin/nrf52dk/key.pem 
2c36b09db52d4bd8abbc8497c6f7863b  bin/nrf52dk/key.pem

# now with deterministic keys. YOU should get the same hash.

$ make clean 
$ IMGTOOL_DETERMINISTIC=1 make mcuboot-create-key 
$ md5sum bin/nrf52dk/key.pem 
b759c318691237c2a2fa40a475e48690  bin/nrf52dk/key.pem
$ make clean 
$ IMGTOOL_DETERMINISTIC=1 make mcuboot-create-key 
$ md5sum bin/nrf52dk/key.pem 
b759c318691237c2a2fa40a475e48690  bin/nrf52dk/key.pem
```

Now we try to sign twice with the same keys:

```sh
$ make 
# ...
# some output
# ...
$ md5sum bin/nrf52dk/signed-tests_mcuboot.bin bin/nrf52dk/key.pem
1b6aea2e823e0ab366272b56f5a56d65  bin/nrf52dk/signed-tests_mcuboot.bin
b759c318691237c2a2fa40a475e48690  bin/nrf52dk/key.pem
# notice I don't clean, I want to use the same key
$ make 
# ...
# some output
# ...
$ md5sum bin/nrf52dk/signed-tests_mcuboot.bin bin/nrf52dk/key.pem
d0743c0694f76251173544b90ab357e4  bin/nrf52dk/signed-tests_mcuboot.bin
b759c318691237c2a2fa40a475e48690  bin/nrf52dk/key.pem


# sign twice with deterministic signing


$ IMGTOOL_DETERMINISTIC=1 make
# ...
# some output
# ...
$ md5sum bin/nrf52dk/signed-tests_mcuboot.bin bin/nrf52dk/key.pem
1c8b92817d82d1b8ed26b0f2ce2e6f3f  bin/nrf52dk/signed-tests_mcuboot.bin
b759c318691237c2a2fa40a475e48690  bin/nrf52dk/key.pem
# notice I don't clean, I want to use the same key, though in this case it should not
# matter because the key is always the same
$ IMGTOOL_DETERMINISTIC=1 make 
# ...
# some output
# ...
$ md5sum bin/nrf52dk/signed-tests_mcuboot.bin bin/nrf52dk/key.pem
1c8b92817d82d1b8ed26b0f2ce2e6f3f  bin/nrf52dk/signed-tests_mcuboot.bin
b759c318691237c2a2fa40a475e48690  bin/nrf52dk/key.pem
```

### Issues/PRs references

Depends on #10612 .
